crates/task-impls/src/request.rs

```diff
@@ -0,0 +1,261 @@
+/// a request.  If at any point the requested info is seen in the data stores or
+/// the view has moved beyond the view we are requesting, the task will completed.
+struct DelayedRequester<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+    /// Network to send requests
+    network: I::QuorumNetwork,
+    /// Shared state to check if the data go populated
+    state: Arc<RwLock<Consensus<TYPES>>>,
+    /// Channel to send the event when we receive a response
+    sender: Sender<HotShotEvent<TYPES>>,
+    /// Duration to delay sending the first request
+    delay: Duration,
+    /// The peers we will request in a random order
+    recipients: Vec<TYPES::SignatureKey>,
+}
+
+/// Wrapper for the info in a VID request
+struct VidRequest<TYPES: NodeType>(TYPES::Time, TYPES::SignatureKey);
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> DelayedRequester<TYPES, I> {
+    /// Wait the delay, then try to complete the request.  Iterates over peers
+    /// until the request is completed, or the data is no longer needed.
+    async fn run(mut self, request: RequestKind<TYPES>, signature: Signature<TYPES>) {
+        // Do the delay then start sending
+        async_sleep(self.delay).await;
+        match request {
+            RequestKind::VID(view, key) => self.do_vid(VidRequest(view, key), signature).await,
+            RequestKind::DAProposal(..) => {}
+        }
+    }
+
+    /// Handle sending a VID Share request, runs the loop until the data exists
+    async fn do_vid(&mut self, req: VidRequest<TYPES>, signature: Signature<TYPES>) {
+        let message = make_vid(&req, signature);
+
+        while !self.recipients.is_empty() && !self.cancel_vid(&req).await {
+            match async_timeout(
+                REQUEST_TIMEOUT,
+                self.network
+                    .request_data::<TYPES>(message.clone(), self.recipients.pop().unwrap()),
+            )
+            .await
+            {
+                Ok(Ok(response)) => {
+                    match response {
+                        ResponseMessage::Found(data) => {
+                            self.handle_response_message(data).await;
+                            // keep trying, but expect the map to be populated, or view to increase
+                            async_sleep(REQUEST_TIMEOUT).await;
+                        }
+                        ResponseMessage::NotFound => {
+                            info!("Peer Responded they did not have the data");
+                        }
+                        ResponseMessage::Denied => {
+                            error!("Request for data was denied by the receiver");
+                        }
+                    }
+                }
+                Ok(Err(e)) => {
+                    warn!("Error Sending request.  Error: {:?}", e);
+                }
+                Err(_) => {
+                    warn!("Request to other node timed out");
```

Comment:
Should these be error logs?

Reply:
I think in the spirit of reducing our error logs warn is ok here.  These errors are hit when we can't reach one specific peer.  It's abnormal, but expected.  The above `error!` log for Denied is error because we know we have stake and a valid key so something is wrong if we get that.